src/Features/LanguageServer/Protocol/Handler/Completion/AbstractLspCompletionResultCreationService.cs

```diff
@@ -132,18 +133,24 @@ async Task<LSP.CompletionItem> CreateLSPCompletionItemAsync(CompletionItem item,
                 lspItem.Kind = GetCompletionKind(item.Tags, capabilityHelper.SupportedItemKinds);
                 lspItem.Preselect = item.Rules.MatchPriority == MatchPriority.Preselect;
 
-                if (!lspItem.Preselect &&
-                    !lspVSClientCapability &&
-                    typedText.Length == 0 &&
-                    item.Rules.SelectionBehavior != CompletionItemSelectionBehavior.HardSelection)
+                if (lspVSClientCapability)
                 {
-                    // VSCode does not have the concept of soft selection, the list is always hard selected.
-                    // In order to emulate soft selection behavior for things like argument completion, regex completion,
-                    // datetime completion, etc. we create a completion item without any specific commit characters.
-                    // This means only tab / enter will commit. VS supports soft selection, so we only do this for non-VS clients.
-                    //
-                    // Note this only applies when user hasn't actually typed anything and completion provider does not request the item
-                    // to be hard-selected. Otherwise, we set its commit characters as normal. This also means we'd need to set IsIncomplete to true
+                    lspItem.CommitCharacters = GetCommitCharacters(item, commitCharactersRuleCache);
+                    return lspItem;
+                }
+
+                // VSCode does not have the concept of soft selection, the list is always hard selected.
```

Comment:
I'm curious about this. It's not just VSCode that doesn't have soft selection, VS doesn't have it either in the LSP client. Should the above block (checking `lspVSClientCapability`) even be present? Certainly we have this issue in VS in Razor, so is it just that nobody turns on LSP for C# files in VS, so this code is incorrect but we're just assuming it's correct?

Reply:
VS client has suggestion mode (`VSInternalCompletionList.SuggestionMode`), so I assumed it has similar behavior s in VS. Maybe I was wrong, will need to try it.

Reply:
> VS doesn't have it either in the LSP client.

Wat.

Reply:
This is the doc-comment for VS client's suggestion mode, which suggests there's soft-selection
```cs
 //
 // Summary:
 //     Gets or sets a value indicating whether the completion list should use suggestion
 //     mode. In suggestion mode items are "soft-selected" by default.
 [DataMember(Name = "_vs_suggestionMode")]
 [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
 public bool SuggestionMode { get; set; }
```

Reply:
TIL. Sounds like we need to try that! Thanks